### PR TITLE
[WIP] feat(pubsub): Add support for JSON decoding in subscriber

### PR DESCRIPTION
**Pull request by opcua-tsn-team-kalycito**
- Add decodeNetworkMessageJson function to decode the received
   networkMessage that is present in JSON encoded format
 - Add receiveBufferedNetworkMessageJson which receives the JSON packets
   and invoke decodeNetworkMessageJSON function to decode the same
 - Add tutorial_pubsub_publish_json.c and tutorial_pubsub_subscribe_json.c
   examples to depict the JSON encoding and decoding use cases

### CODE REVIEW

File: src/pubsub/ua_pubsub_reader.c [hunk 1]
```diff
@@ -29,6 +29,11 @@
 #include "ua_pubsub_bufmalloc.h"
 #endif
 
+#ifdef UA_ENABLE_JSON_ENCODING
+#include "ua_pubsub_networkmessage_json.c"
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:33]: Don't include a .c file, only headers!

File: src/pubsub/ua_pubsub_reader.c [hunk 2]
```diff
@@ -1299,6 +1312,21 @@ decodeNetworkMessage(UA_Server *server, UA_ByteString *buffer, size_t *pos,
...
 
+#ifdef UA_ENABLE_JSON_ENCODING
+UA_StatusCode
+decodeNetworkMessageJson(UA_Server *server, UA_ByteString *src,
+                         UA_NetworkMessage *dst, UA_PubSubConnection *connection) {
+#ifdef UA_DEBUG_DUMP_PKGS
+    UA_dump_hex_pkg(buffer->data, buffer->length);
+#endif
+    UA_StatusCode rv = UA_NetworkMessage_decodeJson(dst,src);
+    UA_CHECK_STATUS_ERROR(rv, return rv,
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_reader.c:1323]: That function does not much besides calling UA_NetworkMessage_decodeJson.
Could be unwrapped inside of decodeAndProcessNetworkMessageJson to reduce the levels of indirection.

File: src/pubsub/ua_pubsub_readergroup.c [hunk 3]
```diff
@@ -656,8 +656,11 @@ UA_ReaderGroup_subscribeCallback(UA_Server *server,
         UA_ReaderGroup_setPubSubState(server, UA_PUBSUBSTATE_ERROR, readerGroup);
         return;
     }
-
+#ifdef UA_ENABLE_JSON_ENCODING
+    receiveBufferedNetworkMessageJson(server, readerGroup, connection);
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_readergroup.c:660]: If `UA_ENABLE_JSON_ENCODING` is enabled, we still need to support the binary encoding.
So add an if/else here that looks at the configured encoding.

File: src/ua_types_encoding_json.c [hunk 4]
```diff
@@ -2561,9 +2561,9 @@ DECODE_JSON(DateTime) {
     
     /* TODO: proper ISO 8601:2004 parsing, musl strptime!*/
     /* DateTime  ISO 8601:2004 without milli is 20 Characters, with millis 24 */
-    if(tokenSize != 20 && tokenSize != 24) {
-        return UA_STATUSCODE_BADDECODINGERROR;
-    }
+    // if(tokenSize != 20 && tokenSize != 24) {
```
Comment (jpfr) [inline src/ua_types_encoding_json.c:2564]: Why this?

File: src/pubsub/ua_pubsub_networkmessage_json.c [hunk 5]
```diff
@@ -505,8 +508,8 @@ static status NetworkMessage_decodeJsonInternal(UA_NetworkMessage *dst, CtxJson
 
     dst->messageIdEnabled = entries[0].found;
     dst->publisherIdEnabled = entries[2].found;
-    if(dst->publisherIdEnabled)
-        dst->publisherIdType = UA_PUBLISHERDATATYPE_STRING;
+    // if(dst->publisherIdEnabled)
```
Comment (jpfr) [inline src/pubsub/ua_pubsub_networkmessage_json.c:511]: Remove the commented lines if they are no longer needed.

**Review by jpfr — COMMENTED** [review]
Looks good!
Please address the comments and the remaining CI issues before merging.

**Comment by andreasebner** [comment]
@opcua-tsn-team-kalycito Is there someone working on this feature? It would be great if we can bring this into master.

**Comment by andreasebner** [comment]
Thank you for the contribution. The work is part of #5234 and merged to the mainline.

**Pull request closed without merging**